hydra-chain-observer/src/Hydra/Blockfrost/ChainObserver.hs

```diff
@@ -0,0 +1,222 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+
+module Hydra.Blockfrost.ChainObserver where
+
+import Hydra.Prelude
+
+import Blockfrost.Client (
+  BlockfrostClientT,
+  runBlockfrost,
+ )
+import Blockfrost.Client qualified as Blockfrost
+import Control.Concurrent.Class.MonadSTM (
+  MonadSTM (readTVarIO),
+  newTVarIO,
+  writeTVar,
+ )
+import Control.Retry (constantDelay, retrying)
+import Hydra.Cardano.Api (
+  BlockHeader,
+  ChainPoint (..),
+  HasTypeProxy (..),
+  Hash,
+  NetworkId (..),
+  NetworkMagic (..),
+  SerialiseAsCBOR (..),
+  SlotNo (..),
+  Tx,
+  UTxO,
+ )
+import Hydra.Chain.Direct.Handlers (convertObservation)
+import Hydra.ChainObserver.NodeClient (
+  ChainObservation (..),
+  ChainObserverLog (..),
+  NodeClient (..),
+  ObserverHandler,
+  logOnChainTx,
+  observeAll,
+ )
+import Hydra.Logging (Tracer, traceWith)
+import Hydra.Tx (IsTx (..))
+
+data APIBlockfrostError
+  = BlockfrostError Text
+  | DecodeError Text
+  | NotEnoughBlockConfirmations Blockfrost.BlockHash
+  | MissingBlockNo Blockfrost.BlockHash
+  | MissingNextBlockHash Blockfrost.BlockHash
+  deriving (Show, Exception)
+
+runBlockfrostM ::
+  (MonadIO m, MonadThrow m) =>
+  Blockfrost.Project ->
+  BlockfrostClientT IO a ->
+  m a
+runBlockfrostM prj action = do
+  result <- liftIO $ runBlockfrost prj action
+  case result of
+    Left err -> throwIO (BlockfrostError $ show err)
+    Right val -> pure val
+
+blockfrostClient ::
+  Tracer IO ChainObserverLog ->
+  FilePath ->
+  Maybe ChainPoint ->
+  Integer ->
+  NodeClient IO
+blockfrostClient tracer projectPath startChainFrom blockConfirmations = do
+  NodeClient
+    { follow = \_ _ observerHandler -> do
+        prj <- Blockfrost.projectFromFile projectPath
+
+        Blockfrost.Block{_blockHash = (Blockfrost.BlockHash genesisBlockHash)} <-
+          runBlockfrostM prj (Blockfrost.getBlock (Left 0))
+
+        Blockfrost.Genesis
+          { _genesisActiveSlotsCoefficient
+          , _genesisSlotLength
+          , _genesisNetworkMagic
+          } <-
+          runBlockfrostM prj Blockfrost.getLedgerGenesis
+
+        let networkId = fromNetworkMagic _genesisNetworkMagic
+        traceWith tracer ConnectingToExternalNode{networkId}
+
+        chainPoint <-
+          case startChainFrom of
+            Just point -> pure point
+            Nothing -> do
+              toChainPoint <$> runBlockfrostM prj Blockfrost.getLatestBlock
+
+        traceWith tracer StartObservingFrom{chainPoint}
+
+        let blockTime = realToFrac _genesisSlotLength / realToFrac _genesisActiveSlotsCoefficient
+
+        let blockHash = fromChainPoint chainPoint genesisBlockHash
+
+        stateTVar <- newTVarIO (blockHash, mempty)
+        void $
+          retrying (retryPolicy blockTime) shouldRetry $ \_ -> do
+            loop tracer prj networkId blockTime observerHandler blockConfirmations stateTVar
+    }
+ where
+  shouldRetry _ = \case
+    Right{} -> pure False
+    Left err -> pure $ isRetryable (spy err)
```

Comment:
we should not use `spy` in production code

Reply:
@ch1bo thought it would be helpful

Reply:
We used this for debugging, yes, but I think it should be removed before merging. It's like `trace`